File: frontend/lib/onboarding/sme/schemas.ts
import { z } from 'zod';
import { cinSchema, isoDateStringSchema } from '@/lib/schemas/company-incorporation';
import { INDIAN_STATES_AND_UTS, REGISTRAR_OF_COMPANIES_OPTIONS } from '@/lib/company-incorporation/options';
import { indianMobileSchema } from '@/lib/auth/schemas';
import {
  AUTHORISED_SIGNATORY_OPTIONS,
  BASIS_OF_AUTHORITY_OPTIONS,
  COMPANY_CLASS_OPTIONS,
  EMPLOYEE_COUNT_RANGE_OPTIONS,
  GST_REGISTRATION_REQUIRED_OPTIONS,
  MERCHANT_BANKER_APPOINTED_OPTIONS,
  PREPARATION_STAGE_OPTIONS,
  PRIMARY_INDUSTRY_OPTIONS,
  PROPOSED_ISSUE_TYPE_OPTIONS,
  RELATIONSHIP_OPTIONS,
  SME_EXCHANGE_OPTIONS,
  TARGET_TIMELINE_OPTIONS,
  YES_NO_OPTIONS,
  YES_NO_UNSURE_OPTIONS,
} from '@/lib/onboarding/sme/constants';

const PAN_REGEX = /^[A-Z]{5}[0-9]{4}[A-Z]$/;
const GSTIN_REGEX = /^[0-9]{2}[A-Z]{5}[0-9]{4}[A-Z][1-9A-Z]Z[0-9A-Z]$/;
const UDYAM_REGEX = /^UDYAM-[A-Z]{2}-[0-9]{2}-[0-9]{7}$/;
const IEC_REGEX = /^([A-Z]{5}[0-9]{4}[A-Z]|[0-9]{10})$/;

const normalizeUppercaseTrim = (value: unknown) =>
  typeof value === 'string' ? value.trim().toUpperCase() : value;

const normalizeTrim = (value: unknown) =>
  typeof value === 'string' ? value.trim() : value;

const requiredSelect = (values: readonly string[], label: string) =>
  z.string().min(1, `${label} is required`).refine((v) => values.includes(v), `${label} is required`);

const indianPinCodeSchema = z
  .string()
  .trim()
  .regex(/^[1-9][0-9]{5}$/, 'PIN code must be a valid 6-digit Indian PIN code');

const notFutureIsoDate = isoDateStringSchema.refine(
  (value) => {
    const parsed = new Date(`${value}T23:59:59.999Z`);
    return !Number.isNaN(parsed.getTime()) && parsed.getTime() <= Date.now();
  },
  { message: 'Date of incorporation cannot be in the future' },
);

const alternateContactSchema = z.object({
  fullName: z.preprocess(normalizeTrim, z.string().min(1, 'Full name is required')),
  designation: z.preprocess(normalizeTrim, z.string().min(1, 'Designation is required')),
  email: z.preprocess(
    (v) => (typeof v === 'string' ? v.trim().toLowerCase() : v),
    z.string().min(1, 'Work email is required').email('Enter a valid email address'),
  ),
  mobile: indianMobileSchema,
});

export const roleAuthorityStepSchema = z
  .object({
    designation: z.preprocess(normalizeTrim, z.string().min(1, 'Designation is required')),
    relationship: requiredSelect(
      RELATIONSHIP_OPTIONS.map((o) => o.value),
      'Relationship with the company',
    ),
    relationshipOther: z.string(),
    authorisedSignatory: requiredSelect(
      AUTHORISED_SIGNATORY_OPTIONS.map((o) => o.value),
      'Authorised signatory',
    ),
    basisOfAuthority: z.string(),
    basisOfAuthorityOther: z.string(),
    primaryOnboardingContact: requiredSelect(
      YES_NO_OPTIONS.map((o) => o.value),
      'Primary onboarding contact',
    ),
    addAlternateContact: z.boolean(),
    alternateContact: z.object({
      fullName: z.string(),
      designation: z.string(),
      email: z.string(),
      mobile: z.string(),
    }),
  })
  .superRefine((data, ctx) => {
    if (data.relationship === 'other' && !data.relationshipOther.trim()) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Please describe the relationship',
        path: ['relationshipOther'],
      });
    }

    const needsBasis =
      data.authorisedSignatory === 'yes' || data.authorisedSignatory === 'unsure';
    if (needsBasis) {
      if (!data.basisOfAuthority) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: 'Basis of authority is required',
          path: ['basisOfAuthority'],
        });
      } else if (
        data.basisOfAuthority === 'other' &&
        !data.basisOfAuthorityOther.trim()
      ) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: 'Please describe the basis of authority',
          path: ['basisOfAuthorityOther'],
        });
      }
    }

    const needsAlternate =
      data.primaryOnboardingContact === 'no' || data.addAlternateContact;
    if (needsAlternate) {
      const result = alternateContactSchema.safeParse(data.alternateContact);
      if (!result.success) {
        result.error.issues.forEach((issue) => {
          ctx.addIssue({ ...issue, path: ['alternateContact', ...issue.path] });
        });
      }
    }
  });

export const registeredOfficeSchema = z.object({
  addressLine1: z.preprocess(normalizeTrim, z.string().min(1, 'Address line 1 is required')),
  addressLine2: z.string(),
  locality: z.string(),
  city: z.preprocess(normalizeTrim, z.string().min(1, 'City is required')),
  district: z.string(),
  state: requiredSelect([...INDIAN_STATES_AND_UTS], 'State or Union Territory'),
  pinCode: indianPinCodeSchema,
  country: z.preprocess(normalizeTrim, z.string().min(1, 'Country is required')),
});

export const companyIdentityStepSchema = z.object({
  legalName: z.preprocess(normalizeTrim, z.string().min(1, 'Legal name is required')),
  cin: cinSchema,
  incorporationDate: notFutureIsoDate,
  companyClass: requiredSelect(
    COMPANY_CLASS_OPTIONS.map((o) => o.value),
    'Company class',
  ),
  registeredState: requiredSelect([...INDIAN_STATES_AND_UTS], 'Registered state or Union Territory'),
  registrarOfCompanies: requiredSelect(
    [...REGISTRAR_OF_COMPANIES_OPTIONS],
    'Registrar of Companies',
  ),
  registeredOffice: registeredOfficeSchema,
  companyEmail: z.preprocess(
    (v) => (typeof v === 'string' ? v.trim().toLowerCase() : v),
    z.string().min(1, 'Company email is required').email('Enter a valid email address'),
  ),
  companyWebsite: z.preprocess(
    (v) => (typeof v === 'string' ? v.trim() : v),
    z.union([z.literal(''), z.string().url('Enter a valid website URL')]),
  ),
});

export const gstRegistrationEntrySchema = z.object({
  id: z.string(),
  gstin: z.preprocess(
    normalizeUppercaseTrim,
    z
      .string()
      .min(1, 'GSTIN is required')
      .regex(GSTIN_REGEX, 'Enter a valid GSTIN in the 15-character format'),
  ),
  state: requiredSelect([...INDIAN_STATES_AND_UTS], 'State or Union Territory'),
  principalPlaceOfBusiness: z.string(),
});

export const businessClassificationStepSchema = z
  .object({
    primaryIndustry: requiredSelect(
      PRIMARY_INDUSTRY_OPTIONS.map((o) => o.value),
      'Primary industry',
    ),
    primaryIndustryOther: z.string(),
    businessSector: z.preprocess(normalizeTrim, z.string().min(1, 'Business sector is required')),
    operationsDescription: z.preprocess(
      normalizeTrim,
      z
        .string()
        .min(30, 'Provide at least 30 characters')
        .max(1000, 'Maximum 1,000 characters'),
    ),
    pan: z.preprocess(
      normalizeUppercaseTrim,
      z
        .string()
        .min(1, 'PAN is required')
        .regex(PAN_REGEX, 'Enter a valid PAN in the AAAAA9999A format'),
    ),
    gstRegistrationRequired: requiredSelect(
      GST_REGISTRATION_REQUIRED_OPTIONS.map((o) => o.value),
      'GST registration requirement',
    ),
    gstRegistrations: z.array(gstRegistrationEntrySchema),
    udyamRegistration: z.string(),
    importExportCode: z.string(),
    employeeCountRange: requiredSelect(
      EMPLOYEE_COUNT_RANGE_OPTIONS.map((o) => o.value),
      'Employee count range',
    ),
  })
  .superRefine((data, ctx) => {
    if (data.primaryIndustry === 'other' && !data.primaryIndustryOther.trim()) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Please describe the primary industry',
        path: ['primaryIndustryOther'],
      });
    }

    if (data.gstRegistrationRequired === 'yes' && data.gstRegistrations.length === 0) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Add at least one GSTIN when GST registration is required',
        path: ['gstRegistrations'],
      });
    }

    if (data.udyamRegistration.trim()) {
      const normalized = data.udyamRegistration.trim().toUpperCase();
      if (!UDYAM_REGEX.test(normalized)) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: 'Enter a valid Udyam registration number in the UDYAM-XX-00-0000000 format',
          path: ['udyamRegistration'],
        });
      }
    }

    if (data.importExportCode.trim()) {
      const normalized = data.importExportCode.trim().toUpperCase();
      if (!IEC_REGEX.test(normalized)) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: 'Enter a valid IEC in PAN format or as a 10-digit numeric code',
          path: ['importExportCode'],
        });
      }
    }
  });

const requiredNonNegativeInt = (label: string) =>
  z
    .string()
    .min(1, `${label} is required`)
    .refine((v) => /^\d+$/.test(v), `${label} must be a whole number`)
    .transform((v) => Number(v))
    .pipe(z.number().int().min(0, `${label} cannot be negative`));

const requiredPositiveInt = (label: string) =>
  z
    .string()
    .min(1, `${label} is required`)
    .refine((v) => /^\d+$/.test(v), `${label} must be a whole number`)
    .transform((v) => Number(v))
    .pipe(z.number().int().min(1, 'At least one director is required'));

const requiredPercent = (label: string) =>
  z
    .string()
    .min(1, `${label} is required`)
    .refine((v) => /^\d+(\.\d+)?$/.test(v), 'Enter a valid percentage')
    .transform((v) => Number(v))
    .pipe(
      z
        .number()
        .min(0, 'Must be between 0 and 100')
        .max(100, 'Must be between 0 and 100'),
    );

export const ownershipSnapshotStepSchema = z
  .object({
    promoterCount: requiredNonNegativeInt('Promoter count'),
    directorCount: requiredPositiveInt('Director count'),
    promoterHoldingPercent: requiredPercent('Promoter holding'),
    nonPromoterHoldingPercent: requiredPercent('Non-promoter holding'),
    institutionalShareholdersPresent: requiredSelect(
      YES_NO_UNSURE_OPTIONS.map((o) => o.value),
      'Institutional shareholders',
    ),
    foreignShareholdersPresent: requiredSelect(
      YES_NO_UNSURE_OPTIONS.map((o) => o.value),
      'Foreign shareholders',
    ),
    promoterGroupEntitiesPresent: requiredSelect(
      YES_NO_UNSURE_OPTIONS.map((o) => o.value),
      'Promoter group entities',
    ),
  })
  .superRefine((data, ctx) => {
    const total = data.promoterHoldingPercent + data.nonPromoterHoldingPercent;
    if (Math.abs(total - 100) > 0.01) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Promoter and non-promoter holding must total 100%',
        path: ['nonPromoterHoldingPercent'],
      });
    }
  });

export const ipoIntentStepSchema = z
  .object({
    proposedIssueType: requiredSelect(
      PROPOSED_ISSUE_TYPE_OPTIONS.map((o) => o.value),
      'Proposed issue type',
    ),
    issueSizeCrore: z.string(),
    issueSizeNotDecided: z.boolean(),
    targetTimeline: requiredSelect(
      TARGET_TIMELINE_OPTIONS.map((o) => o.value),
      'Target timeline',
    ),
    intendedExchange: requiredSelect(
      SME_EXCHANGE_OPTIONS.map((o) => o.value),
      'Intended SME exchange',
    ),
    primaryPurposes: z.array(z.string()).min(1, 'Select at least one purpose'),
    primaryPurposeOther: z.string(),
    merchantBankerAppointed: requiredSelect(
      MERCHANT_BANKER_APPOINTED_OPTIONS.map((o) => o.value),
      'Merchant banker appointment status',
    ),
    merchantBankerName: z.string(),
    preparationStage: requiredSelect(
      PREPARATION_STAGE_OPTIONS.map((o) => o.value),
      'Current preparation stage',
    ),
  })
  .superRefine((data, ctx) => {
    if (data.primaryPurposes.includes('other') && !data.primaryPurposeOther.trim()) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Please describe the other purpose',
        path: ['primaryPurposeOther'],
      });
    }

    if (!data.issueSizeNotDecided) {
      const amount = Number(data.issueSizeCrore);
      if (!data.issueSizeCrore.trim() || Number.isNaN(amount) || amount <= 0) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: 'Enter a positive issue size in crore or select Not Decided',
          path: ['issueSizeCrore'],
        });
      }
    }

    if (data.merchantBankerAppointed === 'yes' && !data.merchantBankerName.trim()) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Merchant banker name is required',
        path: ['merchantBankerName'],
      });
    }
  });

export const initialDocumentsStepSchema = z.object({
  selections: z.record(z.string(), z.any()),
  skippedForNow: z.boolean(),
});

export const submissionConfirmationsSchema = z.object({
  confirmAccuracy: z.boolean().refine((value) => value === true, {
    message: 'This confirmation is required',
  }),
  confirmAuthorised: z.boolean().refine((value) => value === true, {
    message: 'This confirmation is required',
  }),
  confirmVerification: z.boolean().refine((value) => value === true, {
    message: 'This confirmation is required',
  }),
  agreeTerms: z.boolean().refine((value) => value === true, {
    message: 'You must agree to the Terms and Privacy Policy',
  }),
});

export const stepSchemas = {
  1: roleAuthorityStepSchema,
  2: companyIdentityStepSchema,
  3: businessClassificationStepSchema,
  4: ownershipSnapshotStepSchema,
  5: ipoIntentStepSchema,
  6: initialDocumentsStepSchema,
  7: submissionConfirmationsSchema,
} as const;

export type StepSchemaMap = typeof stepSchemas;

File: backend/app/modules/onboarding/sme/schemas.py
import re
from datetime import date, datetime
from typing import Any, Self

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator, model_validator
from pydantic.alias_generators import to_camel

from app.modules.auth.validation import normalize_phone_e164
from app.modules.onboarding.sme.india_constants import (
    INDIAN_STATES_AND_UTS,
    REGISTRAR_OF_COMPANIES_OPTIONS,
)
from app.modules.onboarding.sme.step_mappings import DOCUMENT_IDS

PAN_REGEX = re.compile(r"^[A-Z]{5}[0-9]{4}[A-Z]$")
GSTIN_REGEX = re.compile(r"^[0-9]{2}[A-Z]{5}[0-9]{4}[A-Z][1-9A-Z]Z[0-9A-Z]$")
UDYAM_REGEX = re.compile(r"^UDYAM-[A-Z]{2}-[0-9]{2}-[0-9]{7}$")
IEC_REGEX = re.compile(r"^([A-Z]{5}[0-9]{4}[A-Z]|[0-9]{10})$")
CIN_REGEX = re.compile(r"^[A-Z0-9]{21}$")
PIN_REGEX = re.compile(r"^[1-9][0-9]{5}$")

RELATIONSHIP_OPTIONS = {
    "promoter", "director", "kmp", "employee", "professional-adviser",
    "authorised-external-representative", "other",
}
AUTHORISED_SIGNATORY_OPTIONS = {"yes", "no", "unsure"}
BASIS_OF_AUTHORITY_OPTIONS = {
    "board-resolution", "power-of-attorney", "employment-position",
    "constitutional-authority", "other",
}
YES_NO_OPTIONS = {"yes", "no"}
YES_NO_UNSURE_OPTIONS = {"yes", "no", "unsure"}
COMPANY_CLASS_OPTIONS = {"public", "private"}
GST_REQUIRED_OPTIONS = {"yes", "no", "unsure"}
EMPLOYEE_COUNT_OPTIONS = {
    "1-10", "11-50", "51-100", "101-250", "251-500", "501-1000", "1000-plus",
}
ISSUE_TYPE_OPTIONS = {"fresh-issue", "offer-for-sale", "combination", "not-decided"}
TIMELINE_OPTIONS = {
    "within-3-months", "3-6-months", "6-12-months", "12-18-months",
    "more-than-18-months", "not-decided",
}
EXCHANGE_OPTIONS = {"nse-emerge", "bse-sme", "not-decided"}
MERCHANT_BANKER_OPTIONS = {"yes", "no", "in-discussion", "not-decided"}
PREPARATION_STAGE_OPTIONS = {
    "exploring", "internal-preparation", "advisers-being-appointed",
    "due-diligence-started", "drafting-started", "filing-preparation", "not-sure",
}
PRIMARY_INDUSTRY_OPTIONS = {
    "agriculture-allied", "automotive", "chemicals", "consumer-products",
    "electronics-electrical", "engineering-capital-goods", "financial-services",
    "food-processing", "healthcare-pharma", "it-software", "infrastructure-construction",
    "logistics-transportation", "manufacturing", "media-entertainment",
    "professional-services", "renewable-energy", "retail-ecommerce", "textiles-apparel", "other",
}
ISSUE_PURPOSE_OPTIONS = {
    "capital-expenditure", "working-capital", "debt-repayment", "acquisition",
    "general-corporate-purposes", "offer-for-sale", "brand-building-marketing",
    "technology-investment", "other", "not-decided",
}


class ApiModel(BaseModel):
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
    )


class AlternateContactModel(ApiModel):
    full_name: str = Field(min_length=1)
    designation: str = Field(min_length=1)
    email: EmailStr
    mobile: str

    @field_validator("mobile")
    @classmethod
    def validate_mobile(cls, value: str) -> str:
        return normalize_phone_e164(value)


class RoleAuthorityStepData(ApiModel):
    designation: str = Field(min_length=1)
    relationship: str
    relationship_other: str = ""
    authorised_signatory: str
    basis_of_authority: str = ""
    basis_of_authority_other: str = ""
    primary_onboarding_contact: str
    add_alternate_contact: bool = False
    alternate_contact: AlternateContactModel | dict[str, Any] = Field(default_factory=dict)

    @model_validator(mode="after")
    def validate_role_authority(self) -> Self:
        if self.relationship not in RELATIONSHIP_OPTIONS:
            msg = "Relationship with the company is required"
            raise ValueError(msg)
        if self.relationship == "other" and not self.relationship_other.strip():
            raise ValueError("Please describe the relationship")
        if self.authorised_signatory not in AUTHORISED_SIGNATORY_OPTIONS:
            msg = "Authorised signatory is required"
            raise ValueError(msg)
        if self.authorised_signatory in {"yes", "unsure"}:
            if not self.basis_of_authority:
                raise ValueError("Basis of authority is required")
            if self.basis_of_authority not in BASIS_OF_AUTHORITY_OPTIONS:
                raise ValueError("Basis of authority is required")
            if self.basis_of_authority == "other" and not self.basis_of_authority_other.strip():
                raise ValueError("Please describe the basis of authority")
        if self.primary_onboarding_contact not in YES_NO_OPTIONS:
            raise ValueError("Primary onboarding contact is required")
        needs_alternate = self.primary_onboarding_contact == "no" or self.add_alternate_contact
        if needs_alternate:
            AlternateContactModel.model_validate(self.alternate_contact)
        return self


class RegisteredOfficeModel(ApiModel):
    address_line1: str = Field(min_length=1)
    address_line2: str = ""
    locality: str = ""
    city: str = Field(min_length=1)
    district: str = ""
    state: str
    pin_code: str
    country: str = Field(min_length=1)

    @field_validator("state")
    @classmethod
    def validate_state(cls, value: str) -> str:
        if value not in INDIAN_STATES_AND_UTS:
            msg = "State or Union Territory is required"
            raise ValueError(msg)
        return value

    @field_validator("pin_code")
    @classmethod
    def validate_pin(cls, value: str) -> str:
        if not PIN_REGEX.match(value.strip()):
            raise ValueError("PIN code must be a valid 6-digit Indian PIN code")
        return value.strip()


class CompanyIdentityStepData(ApiModel):
    legal_name: str = Field(min_length=1)
    cin: str
    incorporation_date: str
    company_class: str
    registered_state: str
    registrar_of_companies: str
    registered_office: RegisteredOfficeModel
    company_email: EmailStr
    company_website: str = ""

    @field_validator("cin")
    @classmethod
    def validate_cin(cls, value: str) -> str:
        normalized = value.strip().upper()
        if not CIN_REGEX.match(normalized):
            raise ValueError("CIN must be exactly 21 uppercase alphanumeric characters")
        return normalized

    @field_validator("incorporation_date")
    @classmethod
    def validate_incorporation_date(cls, value: str) -> str:
        parsed = date.fromisoformat(value)
        if parsed > date.today():
            raise ValueError("Date of incorporation cannot be in the future")
        return value

    @field_validator("company_class")
    @classmethod
    def validate_company_class(cls, value: str) -> str:
        if value not in COMPANY_CLASS_OPTIONS:
            raise ValueError("Company class is required")
        return value

    @field_validator("registered_state")
    @classmethod
    def validate_registered_state(cls, value: str) -> str:
        if value not in INDIAN_STATES_AND_UTS:
            raise ValueError("Registered state or Union Territory is required")
        return value

    @field_validator("registrar_of_companies")
    @classmethod
    def validate_roc(cls, value: str) -> str:
        if value not in REGISTRAR_OF_COMPANIES_OPTIONS:
            raise ValueError("Registrar of Companies is required")
        return value

    @field_validator("company_website")
    @classmethod
    def validate_website(cls, value: str) -> str:
        if value and not value.startswith(("http://", "https://")):
            raise ValueError("Enter a valid website URL")
        return value


class GstRegistrationEntryModel(ApiModel):
    id: str
    gstin: str
    state: str
    principal_place_of_business: str = ""

    @field_validator("gstin")
    @classmethod
    def validate_gstin(cls, value: str) -> str:
        normalized = value.strip().upper()
        if not GSTIN_REGEX.match(normalized):
            raise ValueError("Enter a valid GSTIN in the 15-character format")
        return normalized

    @field_validator("state")
    @classmethod
    def validate_state(cls, value: str) -> str:
        if value not in INDIAN_STATES_AND_UTS:
            raise ValueError("State or Union Territory is required")
        return value


class BusinessClassificationStepData(ApiModel):
    primary_industry: str
    primary_industry_other: str = ""
    business_sector: str = Field(min_length=1)
    operations_description: str = Field(min_length=30, max_length=1000)
    pan: str
    gst_registration_required: str
    gst_registrations: list[GstRegistrationEntryModel] = Field(default_factory=list)
    udyam_registration: str = ""
    import_export_code: str = ""
    employee_count_range: str

    @field_validator("primary_industry")
    @classmethod
    def validate_primary_industry(cls, value: str) -> str:
        if value not in PRIMARY_INDUSTRY_OPTIONS:
            raise ValueError("Primary industry is required")
        return value

    @field_validator("pan")
    @classmethod
    def validate_pan(cls, value: str) -> str:
        normalized = value.strip().upper()
        if not PAN_REGEX.match(normalized):
            raise ValueError("Enter a valid PAN in the AAAAA9999A format")
        return normalized

    @field_validator("gst_registration_required")
    @classmethod
    def validate_gst_required(cls, value: str) -> str:
        if value not in GST_REQUIRED_OPTIONS:
            raise ValueError("GST registration requirement is required")
        return value

    @field_validator("employee_count_range")
    @classmethod
    def validate_employee_count(cls, value: str) -> str:
        if value not in EMPLOYEE_COUNT_OPTIONS:
            raise ValueError("Employee count range is required")
        return value

    @model_validator(mode="after")
    def validate_business(self) -> Self:
        if self.primary_industry == "other" and not self.primary_industry_other.strip():
            raise ValueError("Please describe the primary industry")
        if self.gst_registration_required == "yes" and not self.gst_registrations:
            raise ValueError("Add at least one GSTIN when GST registration is required")
        if self.udyam_registration.strip():
            normalized = self.udyam_registration.strip().upper()
            if not UDYAM_REGEX.match(normalized):
                raise ValueError(
                    "Enter a valid Udyam registration number in the UDYAM-XX-00-0000000 format",
                )
        if self.import_export_code.strip():
            normalized = self.import_export_code.strip().upper()
            if not IEC_REGEX.match(normalized):
                raise ValueError("Enter a valid IEC in PAN format or as a 10-digit numeric code")
        return self


class OwnershipSnapshotStepData(ApiModel):
    promoter_count: str
    director_count: str
    promoter_holding_percent: str
    non_promoter_holding_percent: str
    institutional_shareholders_present: str
    foreign_shareholders_present: str
    promoter_group_entities_present: str

    @model_validator(mode="after")
    def validate_ownership(self) -> Self:
        ownership_checks = (
            (
                "Institutional shareholders",
                self.institutional_shareholders_present,
                YES_NO_UNSURE_OPTIONS,
            ),
            (
                "Foreign shareholders",
                self.foreign_shareholders_present,
                YES_NO_UNSURE_OPTIONS,
            ),
            (
                "Promoter group entities",
                self.promoter_group_entities_present,
                YES_NO_UNSURE_OPTIONS,
            ),
        )
        for label, value, options in ownership_checks:
            if value not in options:
                raise ValueError(f"{label} is required")
        if not self.promoter_count.isdigit():
            raise ValueError("Promoter count must be a whole number")
        if not self.director_count.isdigit() or int(self.director_count) < 1:
            raise ValueError("At least one director is required")
        try:
            promoter_pct = float(self.promoter_holding_percent)
            non_promoter_pct = float(self.non_promoter_holding_percent)
        except ValueError as exc:
            raise ValueError("Enter a valid percentage") from exc
        if not 0 <= promoter_pct <= 100 or not 0 <= non_promoter_pct <= 100:
            raise ValueError("Must be between 0 and 100")
        if abs(promoter_pct + non_promoter_pct - 100) > 0.01:
            raise ValueError("Promoter and non-promoter holding must total 100%")
        return self


class IpoIntentStepData(ApiModel):
    proposed_issue_type: str
    issue_size_crore: str = ""
    issue_size_not_decided: bool = False
    target_timeline: str
    intended_exchange: str
    primary_purposes: list[str] = Field(min_length=1)
    primary_purpose_other: str = ""
    merchant_banker_appointed: str
    merchant_banker_name: str = ""
    preparation_stage: str

    @model_validator(mode="after")
    def validate_ipo_intent(self) -> Self:
        if self.proposed_issue_type not in ISSUE_TYPE_OPTIONS:
            raise ValueError("Proposed issue type is required")
        if self.target_timeline not in TIMELINE_OPTIONS:
            raise ValueError("Target timeline is required")
        if self.intended_exchange not in EXCHANGE_OPTIONS:
            raise ValueError("Intended SME exchange is required")
        if self.merchant_banker_appointed not in MERCHANT_BANKER_OPTIONS:
            raise ValueError("Merchant banker appointment status is required")
        if self.preparation_stage not in PREPARATION_STAGE_OPTIONS:
            raise ValueError("Current preparation stage is required")
        if "other" in self.primary_purposes and not self.primary_purpose_other.strip():
            raise ValueError("Please describe the other purpose")
        if not self.issue_size_not_decided:
            try:
                amount = float(self.issue_size_crore)
            except ValueError as exc:
                raise ValueError(
                    "Enter a positive issue size in crore or select Not Decided",
                ) from exc
            if amount <= 0:
                raise ValueError("Enter a positive issue size in crore or select Not Decided")
        if self.merchant_banker_appointed == "yes" and not self.merchant_banker_name.strip():
            raise ValueError("Merchant banker name is required")
        invalid_purposes = set(self.primary_purposes) - ISSUE_PURPOSE_OPTIONS
        if invalid_purposes:
            raise ValueError("Select at least one purpose")
        return self


class DocumentSelectionMeta(ApiModel):
    file_name: str = Field(min_length=1)
    file_size: int = Field(ge=0)
    mime_type: str = Field(min_length=1)


class InitialDocumentsStepData(ApiModel):
    selections: dict[str, DocumentSelectionMeta | None]
    skipped_for_now: bool = False

    @model_validator(mode="after")
    def validate_documents(self) -> Self:
        for document_id in DOCUMENT_IDS:
            self.selections.setdefault(document_id, None)
        for document_id, selection in self.selections.items():
            if document_id not in DOCUMENT_IDS:
                continue
            if selection is None:
                continue
            if isinstance(selection, dict):
                DocumentSelectionMeta.model_validate(selection)
        return self


class SubmissionConfirmationsData(ApiModel):
    confirm_accuracy: bool
    confirm_authorised: bool
    confirm_verification: bool
    agree_terms: bool

    @model_validator(mode="after")
    def validate_confirmations(self) -> Self:
        for field_name, value in (
            ("confirmAccuracy", self.confirm_accuracy),
            ("confirmAuthorised", self.confirm_authorised),
            ("confirmVerification", self.confirm_verification),
            ("agreeTerms", self.agree_terms),
        ):
            if not value:
                raise ValueError(f"{field_name} confirmation is required")
        return self


STEP_MODELS: dict[str, type[ApiModel]] = {
    "role_authority": RoleAuthorityStepData,
    "company_identity": CompanyIdentityStepData,
    "business_classification": BusinessClassificationStepData,
    "ownership_snapshot": OwnershipSnapshotStepData,
    "ipo_intent": IpoIntentStepData,
    "initial_documents": InitialDocumentsStepData,
    "review_submit": SubmissionConfirmationsData,
}


class OnboardingApplicationResponse(ApiModel):
    id: str
    status: str
    current_step: str
    completed_steps: list[str]
    draft_data: dict[str, Any]
    schema_version: int
    version: int
    submitted_at: datetime | None = None


class OnboardingSummaryOnlyResponse(ApiModel):
    id: str
    status: str
    current_step: str
    completed_steps: list[str]


class SubmitOnboardingRequest(ApiModel):
    submission_confirmations: SubmissionConfirmationsData


class SubmitOnboardingResponse(ApiModel):
    id: str
    status: str
    next_action: str = "open_dashboard"
    redirect_to: str = "/projects/demo"
